lapack/testlapack/dorgr2.go

```diff
@@ -0,0 +1,96 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"fmt"
+	"testing"
+
+	"golang.org/x/exp/rand"
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/lapack"
+)
+
+type Dorgr2er interface {
+	Dorgr2(m, n, k int, a []float64, lda int, tau []float64, work []float64)
+
+	Dgerqfer
+}
+
+func Dorgr2Test(t *testing.T, impl Dorgr2er) {
+	rnd := rand.New(rand.NewSource(1))
+	dorgr2Test(t, impl, rnd, 2, 2, 1, 2)
+	for _, k := range []int{0, 1, 2, 5} {
+		for _, m := range []int{k, k + 1, k + 2, k + 4} {
+			for _, n := range []int{m, m + 1, m + 2, m + 4, m + 7} {
+				for _, lda := range []int{n, n + 5} {
+					dorgr2Test(t, impl, rnd, m, n, k, lda)
+				}
+			}
+		}
+	}
+}
+
+func dorgr2Test(t *testing.T, impl Dorgr2er, rnd *rand.Rand, m, n, k, lda int) {
+	const tol = 1e-12
+	name := fmt.Sprintf("m=%v,n=%v,k=%v,lda=%v", m, n, k, lda)
+
+	if lda == 0 {
+		lda = n
+	}
+	// k = min(n, m)
+	a := randomGeneral(m, n, lda, rnd)
+	aCopy := cloneGeneral(a)
+	// Compute the RQ decomposition of A.
+	tau := make([]float64, m)
+	work := make([]float64, 1)
+	impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, -1)
+	work = make([]float64, int(work[0]))
+	impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, len(work))
+
+	// Generate the upper triangular matrix R from subarray A[0:m, n-m:n]
+	r := zeros(k, m, m)
+	for i := 0; i < k; i++ {
```

Comment:
Last k rows.

Reply:
Take a look at the changes, I'm not sure If I changed what you meant. I understood this as "I should be generating upper triangular matrix from last rows of A". Last rows of R are all rows of R as R has k rows

Reply:
Yes, last k rows of A. But you're not extracting the triangle correctly. The upper triangle of the lower right kxk block of A should go into the (lower) right kxk block of R.

Reply:
>  The upper triangle of the lower right kxk block of A should go into the (lower) right kxk block of R.

![peele sweat](https://media1.tenor.com/images/9b131a5e928f9cf7be8d4204ed4f765e/tenor.gif?itemid=9416963)

I think I may have gotten it this time. pushing now